File: leetcode/algorithms/217_contains_duplicate/main.py
from typing import List, Set


class ContainsDuplicate:
    """
    # Hash Set
    #
    # Complexities:
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def containsDuplicate(self, nums: List[int]) -> bool:
        if len(nums) < 2:
            return False

        numsSet: Set = set()

        for num in nums:
            if num in numsSet:
                return True
            else:
                numsSet.add(num)

        return False


    # Solution
    """
    # Solution 1
    #
    # Hash Set
    #
    # Complexities:
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution1(self, nums: List[int]) -> bool:
        seen = set()

        for num in nums:
            if num in seen:
                return True
            seen.add(num)

        return False

    """
    # Solution 2
    #
    # Complexities:
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution2(self, nums: List[int]) -> bool:
        return len(nums) != len(set(nums))

    """
    # Solution 3
    #
    # Complexities:
    #   - Time Complexity: O(N * logᴺ)
    #   - Space Complexity: O(1)
    """
    def solution3(self, nums: List[int]) -> bool:
        nums.sort()

        for i in range(len(nums) - 1):
            if nums[i] == nums[i + 1]:
                return True

        return False

File: leetcode/algorithms/217_contains_duplicate/main.dart
class ContainsDuplicate {
  /**
   * Hash Set
   *
   * Complexities:
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  bool containsDuplicate(List<int> nums) {
    if (nums.length < 2) {
      return false;
    }

    Set<int> numsSet = Set();

    for (int num in nums) {
      if (numsSet.contains(num)) {
        return true;
      } else {
        numsSet.add(num);
      }
    }

    return false;
  }


  // Solution
  /**
   * Solution 1
   *
   * Hash Set
   *
   * Complexities:
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  bool solution1(List<int> nums) {
    final Set<int> seen = {};

    for (final num in nums) {
      if (!seen.add(num)) {
        return true;
      }
    }

    return false;
  }

  /**
   * Solution 2
   *
   * Sort
   *
   * Complexities:
   *   - Time Complexity: O(N * logᴺ)
   *   - Space Complexity: O(1)
   */
  bool solution2(List<int> nums) {
    if (nums.length < 2) {
      return false;
    }

    nums.sort();

    for (int i = 1; i < nums.length; i++) {
      if (nums[i] == nums[i - 1]) {
        return true;
      }
    }

    return false;
  }
}
